gsconnect: init at 20

etu opened the pull request:
###### Motivation for this change
It seems like a nice gnome extension that we all want :wink:

I think this may be the 3rd or 4th known try to package this up for NixOS. And after submitting a ticket and waiting a bit they have refactored a lot of their extension to be more friendly for packaging as documented here: https://github.com/andyholmes/gnome-shell-extension-gsconnect/issues/91#issuecomment-423036690

~~Note, this is *far* from complete. The install phase is still missing and isn't working yet.~~ I just wanted to publish my WIP to reduce overlapping efforts.

cc @jtojnar 

Things to do

- [x] Package nautilus-python
- [ ] Test the nautilus integration
- [x] Figure out what to do about the missing Chromium icon in the settings
- [x] Move hard-coding the program paths upstream
- [ ] Test bluetooth connection
- [x] Open tracking issues for the TODO comments
- [ ] Write some docs (something like [this](https://github.com/NixOS/nixpkgs/pull/46779/files#diff-dd4c714984d1e6e681ffa3f64117bb6b)
    - Needs to be installed using `environment.systemPackage`
    - Firewall settings
    - Native messaging hosts setup

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

etu committed:
gnomeExtensions.gsconnect: init at 20

GrahamcOfBorg commented:
<!--REQUEST_ID=7b45d55e-e9c9-41ea-b774-59b5869a51b2-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47059&attempt_id=fd102f01-b173-4a17-90d8-07190bdabda8)

Attempted: gnomeExtensions.gsconnect

<details><summary>Partial log (click to expand)</summary><p>

```
Installing /build/source/build/org.gnome.Shell.Extensions.GSConnect.service to /nix/store/hss6p24dmzs4m1njf0p5y06mf74zwi6j-gnome-shell-gsconnect-13-alpha/share/dbus-1/services
Installing /build/source/src/nautilus-gsconnect.py to /nix/store/hss6p24dmzs4m1njf0p5y06mf74zwi6j-gnome-shell-gsconnect-13-alpha/share/nautilus-python/extensions
Installing /build/source/build/org.gnome.shell.extensions.gsconnect.json to /nix/store/hss6p24dmzs4m1njf0p5y06mf74zwi6j-gnome-shell-gsconnect-13-alpha/lib/mozilla/native-messaging-hosts
Installing /build/source/data/org.gnome.Shell.Extensions.GSConnect.gschema.xml to /nix/store/hss6p24dmzs4m1njf0p5y06mf74zwi6j-gnome-shell-gsconnect-13-alpha/share/glib-2.0/schemas
Running custom install script '/build/source/meson/nmh.sh etc'
FAILED: meson-install
/nix/store/sb88r9crdhinhkm6vpxwzf5zmij86p8v-python3-3.6.6/bin/python3.6m /nix/store/rr64y8pqq6xgxhv2g8xwzxfpd9zy06i6-meson-0.46.1/bin/meson --internal install /build/source/build/meson-private/install.dat
ninja: build stopped: subcommand failed.
builder for '/nix/store/5s0a0wd3sar9i90iadxc314il7yc3m25-gnome-shell-gsconnect-13-alpha.drv' failed with exit code 1
error: build of '/nix/store/5s0a0wd3sar9i90iadxc314il7yc3m25-gnome-shell-gsconnect-13-alpha.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=7b45d55e-e9c9-41ea-b774-59b5869a51b2-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47059&attempt_id=342b41e8-1e67-496d-a84e-056db716ca43)

Attempted: gnomeExtensions.gsconnect

<details><summary>Partial log (click to expand)</summary><p>

```
Installing /build/source/build/org.gnome.Shell.Extensions.GSConnect.service to /nix/store/sjkzj9vfnm9rzf39zqx0md84wksjzyvj-gnome-shell-gsconnect-13-alpha/share/dbus-1/services
Installing /build/source/src/nautilus-gsconnect.py to /nix/store/sjkzj9vfnm9rzf39zqx0md84wksjzyvj-gnome-shell-gsconnect-13-alpha/share/nautilus-python/extensions
Installing /build/source/build/org.gnome.shell.extensions.gsconnect.json to /nix/store/sjkzj9vfnm9rzf39zqx0md84wksjzyvj-gnome-shell-gsconnect-13-alpha/lib/mozilla/native-messaging-hosts
Installing /build/source/data/org.gnome.Shell.Extensions.GSConnect.gschema.xml to /nix/store/sjkzj9vfnm9rzf39zqx0md84wksjzyvj-gnome-shell-gsconnect-13-alpha/share/glib-2.0/schemas
Running custom install script '/build/source/meson/nmh.sh etc'
FAILED: meson-install
/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/bin/python3.6m /nix/store/0wrv6cips8sfzk04ca67w6vlsgy4lb3z-meson-0.46.1/bin/meson --internal install /build/source/build/meson-private/install.dat
ninja: build stopped: subcommand failed.
builder for '/nix/store/11p773nk46yz0q22ihgj5k1976fyvqjh-gnome-shell-gsconnect-13-alpha.drv' failed with exit code 1
error: build of '/nix/store/11p773nk46yz0q22ihgj5k1976fyvqjh-gnome-shell-gsconnect-13-alpha.drv' failed
```
</p></details>

jtojnar commented on pkgs/desktops/gnome-3/extensions/gsconnect/default.nix:
```diff
@@ -0,0 +1,39 @@
+    description = "KDE Connect implementation for Gnome Shell";
+    license = licenses.gpl2;
+    maintainers = with maintainers; [ etu ];
+    homepage = https://extensions.gnome.org/extension/1319/gsconnect/;
```
The extension page points to https://github.com/andyholmes/gnome-shell-extension-gsconnect/wiki as a homepage.

jtojnar commented on pkgs/desktops/gnome-3/extensions/gsconnect/default.nix:
```diff
@@ -0,0 +1,39 @@
+    sshfs
+  ];
+
+  postPatch = "patchShebangs /build/source/meson/nmh.sh";
```
Just patch `meson/nmh.sh`.

jtojnar commented on pkgs/desktops/gnome-3/extensions/gsconnect/default.nix:
```diff
@@ -0,0 +1,39 @@
+    sha256 = "0dynq5b2vrp54r9k6fpc81qjmr2qv1nbbyn4khrjxv283q7w3h7q";
+  };
+
+  # TODO: Missing "dbus-1"
```
D-Bus is not required – it is only used for obtaining install path, but there is a [more convenient fallback](https://github.com/andyholmes/gnome-shell-extension-gsconnect/blob/dccdfa1a63e33f505e38f7d2667c79ff6addd8b1/meson.build#L90).

GrahamcOfBorg commented:
<!--REQUEST_ID=36c3e835-5dd8-4795-88c8-e04475667cb1-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47059&attempt_id=1ebcec56-ba93-4dc2-b4c8-3bcd2e873934)

Attempted: gnomeExtensions.gsconnect

<details><summary>Partial log (click to expand)</summary><p>

```
Installing /build/source/build/org.gnome.Shell.Extensions.GSConnect.service to /nix/store/87bca56kdpvyz2mwhaaxd895hx2yifir-gnome-shell-gsconnect-13-alpha/share/dbus-1/services
Installing /build/source/src/nautilus-gsconnect.py to /nix/store/87bca56kdpvyz2mwhaaxd895hx2yifir-gnome-shell-gsconnect-13-alpha/share/nautilus-python/extensions
Installing /build/source/build/org.gnome.shell.extensions.gsconnect.json to /nix/store/87bca56kdpvyz2mwhaaxd895hx2yifir-gnome-shell-gsconnect-13-alpha/lib/mozilla/native-messaging-hosts
Installing /build/source/data/org.gnome.Shell.Extensions.GSConnect.gschema.xml to /nix/store/87bca56kdpvyz2mwhaaxd895hx2yifir-gnome-shell-gsconnect-13-alpha/share/glib-2.0/schemas
Running custom install script '/build/source/meson/nmh.sh etc'
FAILED: meson-install
/nix/store/hy65mn4wjswqih75gfr6g4q3xgqdm325-python3-3.6.6/bin/python3.6m /nix/store/0wrv6cips8sfzk04ca67w6vlsgy4lb3z-meson-0.46.1/bin/meson --internal install /build/source/build/meson-private/install.dat
ninja: build stopped: subcommand failed.
builder for '/nix/store/c92k5aqgdcnmy2h1wfsb76fb6xkg7q10-gnome-shell-gsconnect-13-alpha.drv' failed with exit code 1
error: build of '/nix/store/c92k5aqgdcnmy2h1wfsb76fb6xkg7q10-gnome-shell-gsconnect-13-alpha.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=36c3e835-5dd8-4795-88c8-e04475667cb1-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47059&attempt_id=1de5a1e7-a7c6-45c8-b16d-492bd5047186)

Attempted: gnomeExtensions.gsconnect

<details><summary>Partial log (click to expand)</summary><p>

```
Installing /build/source/build/org.gnome.Shell.Extensions.GSConnect.service to /nix/store/dxbgmc735cwwszwa30hl7098sjgzbzcq-gnome-shell-gsconnect-13-alpha/share/dbus-1/services
Installing /build/source/src/nautilus-gsconnect.py to /nix/store/dxbgmc735cwwszwa30hl7098sjgzbzcq-gnome-shell-gsconnect-13-alpha/share/nautilus-python/extensions
Installing /build/source/build/org.gnome.shell.extensions.gsconnect.json to /nix/store/dxbgmc735cwwszwa30hl7098sjgzbzcq-gnome-shell-gsconnect-13-alpha/lib/mozilla/native-messaging-hosts
Installing /build/source/data/org.gnome.Shell.Extensions.GSConnect.gschema.xml to /nix/store/dxbgmc735cwwszwa30hl7098sjgzbzcq-gnome-shell-gsconnect-13-alpha/share/glib-2.0/schemas
Running custom install script '/build/source/meson/nmh.sh etc'
FAILED: meson-install
/nix/store/sb88r9crdhinhkm6vpxwzf5zmij86p8v-python3-3.6.6/bin/python3.6m /nix/store/rr64y8pqq6xgxhv2g8xwzxfpd9zy06i6-meson-0.46.1/bin/meson --internal install /build/source/build/meson-private/install.dat
ninja: build stopped: subcommand failed.
builder for '/nix/store/2rjfbd374hvrvw1cbam6xc6avvd1r436-gnome-shell-gsconnect-13-alpha.drv' failed with exit code 1
error: build of '/nix/store/2rjfbd374hvrvw1cbam6xc6avvd1r436-gnome-shell-gsconnect-13-alpha.drv' failed
```
</p></details>

jtojnar commented on pkgs/desktops/gnome-3/extensions/gsconnect/default.nix:
```diff
@@ -0,0 +1,77 @@
+    (substituteAll {
+      src = ./fix-paths.patch;
+      inherit sshfs openssl which;
+      python3 = python3.withPackages [ python3.pkgs.pygobject3 ];
```
We will need to rely on shebang and running the [script](https://github.com/andyholmes/gnome-shell-extension-gsconnect/blob/e4be2356815cb9863f0c02776dae3c7235e2f3a5/src/service/components/folks.py) directly, instead of through [`python3`](https://github.com/andyholmes/gnome-shell-extension-gsconnect/blob/e4be2356815cb9863f0c02776dae3c7235e2f3a5/src/service/components/contacts.js#L285) to be able to use `wrapGAppsHook` on it. The [direct library use](https://github.com/andyholmes/gnome-shell-extension-gsconnect/blob/e4be2356815cb9863f0c02776dae3c7235e2f3a5/src/service/components/folks.py#L18) is also concerning.

etu replied:
This line fails on my system:
```
error: attempt to call something which is not a function but a list, at /home/etu/code/nixpkgs/pkgs/development/interpreters/python/with-packages.nix:3:19
(use '--show-trace' to show detailed location information)
```

jtojnar replied:
Should be `python3.withPackages (pkgs: [ python3.pkgs.pygobject3 ])` but it is still broken until we fix it as described above.

etu commented:
I have actually got the extension to run, not work though. In my gnome session. With only the python change above. It shows up in the panel and conflicts on ports with kdeconnect. So I disabled kdeconnect for now.

And I managed to list my device in kdeconnect on my phone.

When I try to pair kdeconnect with gsconnect I get the following message in the log:
```
Sep 22 22:55:18 ford .daemon.js-wrap[1439]: JS ERROR: Error: Requiring Atspi, version 2.0: Typelib file for namespace 'Atspi', version '2.0' not found
                                            @/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/plugins/mousepad.js:3:7
                                            _handleIdentity@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/device.js:270:17
                                            _init@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/device.js:110:13
                                            _ensureDevice@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/.daemon.js-wrapped:244:22
                                            _onIncomingIdentity@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/lan.js:239:36
                                            @/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/.daemon.js-wrapped:888:2
Sep 22 22:55:19 ford .daemon.js-wrap[1439]: JS ERROR: Error: Requiring Atspi, version 2.0: Typelib file for namespace 'Atspi', version '2.0' not found
                                            @/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/plugins/mousepad.js:3:7
                                            _handleIdentity@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/device.js:270:17
                                            _init@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/device.js:110:13
                                            _ensureDevice@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/.daemon.js-wrapped:244:22
                                            _onIncomingIdentity@/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/lan.js:239:36
                                            @/nix/store/i15vmjnh9m0qqwgdlqnq606q2519qhyx-gnome-shell-gsconnect-20180922.1/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/.daemon.js-wrapped:888:2
Sep 22 22:55:22 ford .gnome-shell-wr[1058]: Object Shell.GenericContainer (0x518ce20), has been already finalized. Impossible to get any property from it.
```

So somethings missing for the JS.

jtojnar commented:
That is pretty weird:

```ShellSession
$ rg shared-library (nix-build -A gnome3.folks)/share/gir-1.0/Folks-0.6.gir
$ rg shared-library (nix-build -A atk.dev)/share/gir-1.0/Atk-1.0.gir
14:             shared-library="/nix/store/qhx7xr563xwlabxvb2azrllj5xdm3zyh-atk-2.28.1/lib/libatk-1.0.so.0"
```

Folks does not declare a library name in the GIR file.

cc @hedning

etu commented:
Nice fixes, now I have it running on a second box and manage to discover and connect my Phone. I can send commands to my phone like `find my phone` stuff from the UI. Running commands from the phone on the gnome desktop seems to work as well.

I tested sending a file from my phone with kdeconnect, that worked and a notification showed up in gnome giving me the option to open the directory or open the file.

So feature-wise, I think we're getting there.

Though it shows as disconnected in the menu all the time, don't know if that's a bug in the package or in the packaging:
![image](https://user-images.githubusercontent.com/461970/45924725-451ced00-bf07-11e8-9b9e-480a4db55d56.png)

etu commented:
This should be the nautilus-python that we're missing: https://wiki.gnome.org/Projects/NautilusPython/ https://gitlab.gnome.org/GNOME/nautilus-python/

jtojnar commented:
Does the libfolks integration work for you with the `LD_LIBRARY_PATH` hack? The need to restart the session is pretty annoying to testing.

From what I have seen before that, everything seems to work, including the connection status in the menu. I am connecting through Wi-Fi with the following firewall settings:

```nix
  networking.firewall.allowedTCPPortRanges = [ { from = 1714; to = 1764; }];
  networking.firewall.allowedUDPPortRanges = [ { from = 1714; to = 1764; }];
```

etu commented:
> Does the libfolks integration work for you with the `LD_LIBRARY_PATH` hack? The need to restart the session is pretty annoying to testing.

You should be able to do Alt+F2 and enter the letter `r` and press enter. This reloads the gnome session. Only works on X11 though.

> From what I have seen before that, everything seems to work, including the connection status in the menu. I am connecting through Wi-Fi with the following firewall settings:

The connection status works for me as well now, don't know why it didn't before.

> ```nix
>   networking.firewall.allowedTCPPortRanges = [ { from = 1714; to = 1764; }];
>   networking.firewall.allowedUDPPortRanges = [ { from = 1714; to = 1764; }];
> ```

I have those in place since before because I used kdeconnect in gnome for clipboard-sync etc.

jtojnar commented:
> I've installed the firefox extension gsconnect and that doesn't connect either.

Did you try `(pkgs.firefox.override { extraNativeMessagingHosts = [ pkgs.gnomeExtensions.gsconnect ]; })`?

> What if you reload the shell and then open the GSconnect settings, go to Service and press the `Restart Service` button?

Will try that.

etu commented:
> > I've installed the firefox extension gsconnect and that doesn't connect either.
> 
> Did you try `(pkgs.firefox.override { extraNativeMessagingHosts = [ pkgs.gnomeExtensions.gsconnect ]; })`?

Nope, I have not. But now I have and it seems to work. Thanks :)

jtojnar commented:
What if you reload the shell and then open the GSconnect settings, go to Service and press the `Restart Service` button?

Nope, does not work. The old one is still running under `dbus.service` according to `systemctl --user status`.

jtojnar commented:
Caribou was missing as well, I added it.

Also reported some errors: https://github.com/andyholmes/gnome-shell-extension-gsconnect/issues/202

andyholmes commented:
FYI, Caribou will probably be removed eventually, since it's deprecated by Gnome. As far as the native messaging host it should exit when the browser exits, since that's the parent process.

etu commented:
@andyholmes I don't know if this is a packaging issue or just a bug.

I'm seeing this happening from time to time:

![image](https://user-images.githubusercontent.com/461970/46097057-20ec3500-c1c1-11e8-9f03-a15051f2237b.png)

andyholmes commented:
Are there any errors in journalctl for gnome-shell? That seems like the extension is being loaded multiple times or something is failing during the `disable()` hook.

etu commented:
@andyholmes I'm sorry, I have seen it happen more than once. But I haven't found a reliable way to reproduce it. But I will look into that this evening (about 9-10 hours from now in my timezone and schedule) to see what I can gather for you :slightly_smiling_face:

andyholmes commented:
@etu I think I just saw the same thing happen, but I didn't have my log open and couldn't find the cause :(

I guess I can confirm it as bug, but if it happens again I'd appreciate any logs you might have, since it seems pretty rare.

etu commented:
@andyholmes Hey man, my laptop has been suspended since the morning. And when I unsuspended it and unlocked gnome I saw some flashing of my dash-to-dock (indicating some reloading in gnome I guess) and then a bunch of icons appeared from gsconnect.

Here you have *everything* in my journalctl since just before I suspended until right after I noticed this and dumped the log: https://ptpb.pw/MzuX/text

andyholmes commented:
@etu this sees to be happening for other testers now, too. There's a bug now [#216](https://github.com/andyholmes/gnome-shell-extension-gsconnect/issues/216) to track it.

etu commented:
@andyholmes Interesting, thanks for the update.

@jtojnar I've squashed all the changes together that we have so far (since it's only one package (so far) so one commit is fine) but also updated to https://github.com/andyholmes/gnome-shell-extension-gsconnect/commit/b66d978776b884fbb3c1586b3cb9832d1586caf1 (latest master) to test if that fixes the encountered bug.

I've also rebased on master since that was a fair bit ahead :)

etu committed:
gnome3.nautilus-python: init at 1.2.2

jtojnar commented on pkgs/desktops/gnome-3/misc/nautilus-python/default.nix:
```diff
@@ -0,0 +1,47 @@
+    domain = "gitlab.gnome.org";
+    owner = "GNOME";
+    repo = "nautilus-python";
+    rev = "44224fac32d2d5284956b55ecd884fdfcb1bec4e";
```
You could use `NAUTILUS_PYTHON_${stdenv.lib.replaceChars ["."] ["_"] version}` but it is better to use `fetchurl` with `mirror://gnome`:

https://github.com/NixOS/nixpkgs/blob/c3a1c0008baf467886806cd00430de7a9fe89d06/pkgs/desktops/gnome-3/core/nautilus/default.nix#L11-L14

jtojnar commented on pkgs/desktops/gnome-3/misc/nautilus-python/default.nix:
```diff
@@ -0,0 +1,47 @@
+  };
+
+  nativeBuildInputs = [
+    automake autoconf libtool pkgconfig which gnome-common gtk-doc
```
Fetching from gnome mirror will also allow us to get rid of autotools stuff.

jtojnar commented on pkgs/desktops/gnome-3/misc/nautilus-python/default.nix:
```diff
@@ -0,0 +1,47 @@
+  # checking for libraries required to embed python... no
+  # configure: error: could not find Python lib
+  configurePhase = ''
+    NOCONFIGURE=1 ./autogen.sh
```
This should generally go to `preConfigure`, though it is not needed with the dist tarball.

jtojnar commented on pkgs/desktops/gnome-3/misc/nautilus-python/default.nix:
```diff
@@ -0,0 +1,47 @@
+    substituteInPlace ./configure \
+        --replace /usr/bin/file "${file}/bin/file"
+
+    ./configure --prefix=$out --sysconfdir=/etc --localstatedir=/var \
```
You can use `configureFlags` if you do not override the `configurePhase`

jtojnar commented on pkgs/desktops/gnome-3/misc/nautilus-python/default.nix:
```diff
@@ -0,0 +1,47 @@
+
+  meta = with stdenv.lib; {
+    homepage = https://wiki.gnome.org/Projects/NautilusPython;
+    license = licenses.gpl2;
```
It is `gpl2Plus`.

jtojnar commented on pkgs/desktops/gnome-3/misc/nautilus-python/default.nix:
```diff
@@ -0,0 +1,47 @@
+    homepage = https://wiki.gnome.org/Projects/NautilusPython;
+    license = licenses.gpl2;
+    description = "Python bindings for the Nautilus Extension API";
+  };
```
Could you add `platforms` and `maintainers` attributes?

jtojnar commented:
Finished packaging `nautilus-python`. Did not include `configureFlags = [ "--enable-gtk-doc" ]` because it fails:

```
Traceback (most recent call last):
  File "/nix/store/mr6ssbhb4a946s2rscx9zqr5lhh79qa6-gtk-doc-1.28/bin/gtkdoc-mkhtml", line 51, in <module>
    sys.exit(mkhtml.run(options))
  File "/nix/store/mr6ssbhb4a946s2rscx9zqr5lhh79qa6-gtk-doc-1.28/share/gtk-doc/python/gtkdoc/mkhtml.py", line 101, in run
    shutil.copy(f, '.')
  File "/nix/store/2brlr94ahy3a9mvcjy0qbqpv8zrb7b7s-python-2.7.15/lib/python2.7/shutil.py", line 133, in copy
    copyfile(src, dst)
  File "/nix/store/2brlr94ahy3a9mvcjy0qbqpv8zrb7b7s-python-2.7.15/lib/python2.7/shutil.py", line 97, in copyfile
    with open(dst, 'wb') as fdst:
IOError: [Errno 13] Permission denied: './right.png'
make[2]: *** [Makefile:620: index.sgml] Error 1
make[2]: Leaving directory '/build/nautilus-python-1.2.2/docs'
make[1]: *** [Makefile:490: all-recursive] Error 1
make[1]: Leaving directory '/build/nautilus-python-1.2.2'
make: *** [Makefile:397: all] Error 2
```

jtojnar commented:
This should hopefully fix the mutter issue.

About the `nautilus-python`, we probably need something like https://github.com/NixOS/nixpkgs/blob/1de9c8a023a4000e250e07f18d42b900c6db6cf8/pkgs/desktops/gnome-3/core/nautilus/extension_dir.patch and https://github.com/NixOS/nixpkgs/blob/af7c57232be7bfdc549cde5242a62eda052d39b4/nixos/modules/services/x11/desktop-managers/gnome3.nix#L163 for https://gitlab.gnome.org/GNOME/nautilus-python/blob/44224fac32d2d5284956b55ecd884fdfcb1bec4e/src/nautilus-python.c#L243

andyholmes commented:
@jtojnar 

Hey, I was going to add a meson option to exclude PackageKit for the next release. Do you still want that or do you think it's fine with the fallback it does?

jtojnar commented:
What is the fallback again? I imagine not displaying the install buttons when the interface is not available?

andyholmes commented:
Basically this, although I think I'll either remove the question marks or maybe make some simple functions that try-catch imports:

![screenshot from 2018-11-18 17-51-05](https://user-images.githubusercontent.com/1265208/48681951-9db0e500-eb5a-11e8-93a5-424dae6626fa.png)

andyholmes commented:
Nevermind about PackageKit; I removed it because it was terrible on every distribution :stuck_out_tongue_closed_eyes:

andyholmes commented:
@jtojnar 

I will be adding support for sftp mounts via Gio sometime soon-ish, but at first this will only be a fallback if `sshfs` is not available. Hopefully I can replace the dependency on `sshfs` if all goes well.

To support private key authentication I'll have to use `ssh-add`, but I'll add a meson option and use the same method as other external programs. Let me know if interacting with `ssh-agent` in this way is a problem and I can leave `sshfs` as a fallback.

andyholmes commented:
@jtojnar added `ssh-add` and `ssh-keygen` to meson options, as with the other binaries. `sshfs` still takes precedence over GVFS and these tools though.

jtojnar commented:
@etu, apparently `nautilus-python` loads the scripts from `XDG_DATA_DIRS` so no need for change there (other than adding the `share/nautilus-python/extensions` directory to `pathsToLink`). The bigger issue is that the python library path is wrong. Maybe we need to build with meson instead of the broken m4 clutter.

```
Jan 07 03:22:34 nixos dbus-daemon[944]: [session uid=1000 pid=944] Successfully activated service 'org.gnome.Nautilus'
Jan 07 03:22:34 nixos .nautilus-wrapp[1400]: g_module_open libpython failed: /nix/store/4ag8gng51y7yw6nr3imwyppqfci3j643-python3-3.7.2/lib64/libpython3.7m.so.1.0: cannot open shared object file: No such file or directory
Jan 07 03:22:34 nixos .nautilus-wrapp[1400]: pygobject initialization failed
Jan 07 03:22:34 nixos .nautilus-wrapp[1400]: nautilus_python_init_python failed
Jan 07 03:22:34 nixos org.gnome.Nautilus[944]: ImportError: could not import gobject (error was: ImportError('/nix/store/gc2y0wx51is1dv4jcz1wshfklj3075rc-python3-3.7.2-env/lib/python3.7/site-packages/gi/_gi.cpython-37m-x86_64-linux-gnu.so: undefined symbol: PyExc_StopIteration'))
```

@andyholmes 20-rc2 seems to work fine in the VM, will need someone to test it with real device.

jtojnar committed:
gnome3: link nautilus-python paths to environment

hedning commented:
I'm testing on hardware now (`20-rc2` on wayland). Looking mostly good so far. Two issues:

- [ ] Toggling plugins in `device -> Advanced` will cause a crash. It looks like the `mousepad` plugin will try to load which fails on wayland:
```
jan. 07 13:23:31 x1 org.gnome.Shell.Extensions.GSConnect[1596]: [/service/device.js:loadPlugin:760]: motorola: "loading 'mousepad' plugin"
jan. 07 13:23:31 x1 org.gnome.Shell.desktop[4963]: syntax error: line 1 of stdin
jan. 07 13:23:31 x1 org.gnome.Shell.desktop[4963]: Errors encountered in stdin; not compiled.
jan. 07 13:23:31 x1 kernel: .daemon.js-wrap[1596]: segfault at c ip 00007f52d122f29f sp 00007ffd32927140 error 4 in libcaribou.so.0.0.0[7f52d1223000+20000]
```

- [ ] The service will log all communications to the journal. This includes clipboard sync, which isn't ideal. Debugging like this should probably be turned off by default.

andyholmes commented:
> It looks like the mousepad plugin will try to load which fails on wayland:

Mousepad should be preempted on Wayland as long as the `XDG_SESSION_TYPE` env variable is set properly.

> The service will log all communications to the journal.

Debugging is only on during the period [*Generate Support Log*](https://github.com/andyholmes/gnome-shell-extension-gsconnect/wiki/Help#generate-support-log) is active or if turned manually in the hidden development panel.

hedning commented:
Huh, the logging is gone after having used the generate support log once (I might've had debugging on from an earlier install or something :man_shrugging: ).

> Mousepad should be preempted on Wayland as long as the XDG_SESSION_TYPE env variable is set properly.

Yeah, I noticed that there were guards in the code for this, so seeing `loading 'mousepad' plugin` was surprising. Managed to catch it in a support log:
```
GSConnect Version: 19
GSConnect Install: system
GJS: 15403
XDG_SESSION_TYPE: wayland
GDMSESSION: gnome
--------------------------------------------------------------------------------
-- Logs begin at Fri 2018-02-16 15:12:25 CET, end at Tue 2019-01-08 06:52:32 CET. --
jan. 08 06:52:03 org.gnome.Shell.Extensions.GSConnect[26154]: [/service/device.js:loadPlugin:760]: motorola: "loading 'mousepad' plugin"
jan. 08 06:52:03 org.gnome.Shell.desktop[4963]: syntax error: line 1 of stdin
jan. 08 06:52:03 org.gnome.Shell.desktop[4963]: Errors encountered in stdin; not compiled.
jan. 08 06:52:03 kernel: show_signal_msg: 8 callbacks suppressed
jan. 08 06:52:03 kernel: .daemon.js-wrap[26154]: segfault at c ip 00007f72d2e2529f sp 00007fffa3dc9a00 error 4 in libcaribou.so.0.0.0[7f72d2e19000+20000]
```

hedning commented:
Been using this for a few days, I think it's ready for merge TBH. There's some really nice improvements over using kdeconnectd, eg. proper notification icons.

hedning commented:
Just to be sure.

@GrahamcOfBorg build gnomeExtensions.gsconnect
